### DISCUSSION TITLE
How to configure nested resources

### DISCUSSION THREAD

**Discussion by bendnorman** [dagster-io/dagster#18181] [Q&A] [answered]
In PUDL, we have a [collection of nested pydantic models](https://github.com/catalyst-cooperative/pudl/blob/dev/src/pudl/settings.py) that we've used to validate our ETL configuration.  We made the parent pydantic model a dagster resource because assets need access the "global" configuration. For example, asset A needs the configuration information for asset A and B.  We configured this resource by [recursively translating](https://github.com/catalyst-cooperative/pudl/blob/d9209f3d9ec876d93d7585f29c9981297d5b4092/src/pudl/settings.py#L677-L709) the pydnatic models to legacy dagster config types. This works but feels janky and I think all of our pydantic settings classes can be converted to the new pydantic backed `ConfigurableResource` class.

I've got a toy example of creating nested `ConfigurableResource` classes that I'm able to configure via python:

```python
from dagster import (
    asset,
    ConfigurableResource,
    Definitions,
)
from pydantic import validator


class SettingsBase(ConfigurableResource):
    years: list[int]

    @validator("years")
    def name_must_contain_space(cls, years):
        working_years = [2021, 2022]
        partitions_not_working = list(set(years) - set(working_years))
        if partitions_not_working:
            raise ValueError("invalid year")
        return years


class EpaCemsSettings(SettingsBase):
    states: list[str]

    @validator("states")
    def name_must_contain_space(cls, states):
        working_states = ["AK"]
        partitions_not_working = list(set(states) - set(working_states))
        if partitions_not_working:
            raise ValueError("invalid state")
        return states


class DatasetsSettings(ConfigurableResource):
    epacems: EpaCemsSettings
    # ... other datasource settings classes that subclass SettingsBase
    dummy_attr: str  # dummy attribute


@asset
def epacems_asset(datasets_settings: DatasetsSettings):
    years = datasets_settings.epacems.years


defs = Definitions(
    [epacems_asset],
    resources={"datasets_settings": DatasetsSettings.configure_at_launch()},
)
```

However, I cannot configure the `datasets_settings` resource via Launchpad. I can change the value of `dummy_attr` but `epacems` does not appear as an option. Is it possible to configure nested `ConfigurableResource`s via the Launchpad and YAML?

**Comment by bendnorman** [comment, marked as answer]
Oh looks like the way to do this is by [also making the nested resources configurable at run time](https://docs.dagster.io/concepts/resources#resources-that-depend-on-other-resources).

**Reply by benpankow** [reply]
Yes, that's right - this is so that nested resources don't end up with deeply nested config values in the launchpad, and also to let you re-use a nested resource among multiple different resources, e.g.

```python
epacems = EpaCemsSettings.configure_at_launch()

defs = Definitions(
    [epacems_asset],
    resources={
        "datasets_settings_1": DatasetsSettings.configure_at_launch(epacems=epacems),
        "datasets_settings_2": DatasetsSettings.configure_at_launch(epacems=epacems),
        "epacems": epacems,
    },
)
